codebase/src/compiler/checker-inference.md

```diff
@@ -0,0 +1,245 @@
+# Type Inference
+
+Typescript has a number of related techniques which together are
+called type inference: places where a type is discovered from
+inspecting values instead of a type annotation. This document
+covers them all in one place even though they're all fairly different.
+
+One thing that that is true of all type inference in Typescript:
+type inference is a separate step that happens before checking. The
+checker will infer a type for a location; then it will check the type
+in the normal way, as if the type had been explicitly written. This
+results in redundant checking when the type inference is simple.
+
+None of these techniques are Hindley-Milner type inference. Instead,
+Typescript adds a few ad-hoc inference techniques to its normal
+type-checking. The result is a system that can infer from many useful
+locations, but nowhere near all of them.
+
+## Initialiser inference
+
+The simplest kind of inference is from initialisers. This inference is
+so simple that I don't believe it has been given a separate name until
+now.
+
+You can see this anywhere a variable, parameter or property has an
+initialiser:
+
+```ts
+let x = 123
+function f(x = 123) {
+}
+class C {
+  x = 123
+}
+```
+
+Remember, inference precedes checking, so checking `let x = 123`
+looks like this:
+
+1. Look for the type of `x`.
+2. There is no annotation, so use the (widened) type of the initialiser: `number`.
+3. Check that the initialiser's type `123` is assignable to `number`.
+
+## Contextual typing
+
+Contextual typing looks upward in the tree for a type based on a type
+annotation. This is unlike initialiser inference, which looks at a *sibling*
+node for a type based on a *value*. For example, in
+
+``` ts
+const f: Callback = (a,b) => a.length + b
+```
+
+The parameters `a` and `b` are contextually typed by the type
+`Callback`. The checker discovers this by looking at the parent nodes
+of `a` and `b` until it finds a type annotation on a variable declaration.
+
+In fact, contextual typing only applies to two kinds of things:
+parameters and literals. But it may find a type in a variety of places.
```

Comment:
I guess jsx literals fall into the "literals" bucket, though maybe not in the same obvious way. :D